src/SlackHookHandler.ts

```diff
@@ -0,0 +1,348 @@
+            return;
+        }
+
+        const text = params.text as string;
+        if (!text) {
+            // TODO(paul): When I started looking at this code there was no lookupAndSendMessage()
+            //   I wonder if this code path never gets called...?
+            // lookupAndSendMessage(params.channel_id, params.timestamp, intent, roomID, token);
+            return;
+        }
+
+        let msg = await this.lookupMessage(params.channel_id as string, params.timestamp as string, token);
+
+        if (!msg) {
+            msg = params;
+        }
+
+        // Restore the original parameters, because we've forgotten a lot of
+        // them by now
+        PRESERVE_KEYS.forEach((k) => msg[k] = params[k]);
+        msg.text = await this.doChannelUserReplacements(msg, text, token);
+        return room.onSlackMessage(msg);
+    }
+
+    private async handleAuthorize(roomOrToken: BridgedRoom|string, params: {[key: string]: string|string[]}) {
+        const oauth2 = this.main.oauth2;
+        if (!oauth2) {
+            log.warn("Wasn't expecting to receive /authorize without OAuth2 configured");
+            return {
+                code: 500,
+                html: `OAuth is not configured on this bridge.`,
+            };
+        }
+        let room: BridgedRoom|null = null;
+        let user: string|null = null;
+        if (typeof roomOrToken === "string") {
+            user = oauth2.getUserIdForPreauthToken(roomOrToken);
+            // This might be a user token.
+            if (!user) {
+                return {
+                    code: 500,
+                    html: "Token not known.",
+                };
+            }
+        } else {
+            room = roomOrToken;
+        }
+
+        log.debug("Exchanging temporary code for full OAuth2 token " +
+            (user ? user : room!.InboundId),
+        );
+
+        try {
+            const result = await oauth2.exchangeCodeForToken(
+                params.code as string,
+                roomOrToken,
+            );
+            log.debug("Got a full OAuth2 token");
+            if (room) { // Legacy webhook
+                room.updateAccessToken(result.access_token, new Set(result.access_scopes));
+                this.main.putRoomToStore(room);
+            } else if (user) { // New event api
+                await this.main.setUserAccessToken(
+                    user,
+                    result.team_id,
+                    result.user_id,
+                    result.access_token,
+                );
+                this.main.updateTeamBotStore(
+                    result.team_id,
+                    result.team_name,
+                    result.bot.bot_user_id,
+                    result.bot.bot_access_token,
+                );
+            }
+        } catch (err) {
+            log.error("Error during handling of an oauth token:", err);
+            return {
+                code: 403,
+                html: `
+<h2>Integration Failed</h2>
```

Comment:
We should not be using this kind of multiline string because it is atrociously bad at indentation. Use string concatenation over multiple lines instead.

Applies to other locations in the diff too.